db_test.go

```diff
@@ -0,0 +1,350 @@
+		if err := tc.query(tc.have, tc.options(tc.optionsArgs)); err != nil {
+			t.Error(err)
+		}
+
+		if !reflect.DeepEqual(tc.have, tc.want) {
+			t.Errorf("Error: expected %v, found %v", tc.want, tc.have)
+		}
+	}
+}
+
+func TestFindQuery(t *testing.T) {
+	type queryFunc func(out interface{}, opts ...fury.QueryOption) error
+	type optionFunc func(conditions interface{}) fury.QueryOption
+
+	cases := []struct {
+		have        interface{}
+		options     optionFunc
+		optionsArgs interface{}
+		query       queryFunc
+		want        interface{}
+	}{
+		{
+			&Account{
+				UserID: 1,
+			},
+			fury.Where,
+			fury.IsEqualsTo("username", "test1"),
+			db.Find,
+			&Account{
+				UserID:    1,
+				Username:  "test1",
+				Password:  "test1",
+				Email:     "test1@test.com",
+				CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+			},
+		},
+		{
+			&Account{},
+			fury.Where,
+			fury.IsGreaterThan("userid", 0),
+			db.Find,
+			&Account{
+				UserID:    1,
+				Username:  "test1",
+				Password:  "test1",
+				Email:     "test1@test.com",
+				CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+			},
+		},
+		{
+			&[]*Account{},
+			fury.Where,
+			fury.IsGreaterThan("account.userid", 0),
+			db.Find,
+			&[]*Account{
+				&Account{
+					UserID:    1,
+					Username:  "test1",
+					Password:  "test1",
+					Email:     "test1@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+				&Account{
+					UserID:    2,
+					Username:  "test2",
+					Password:  "test2",
+					Email:     "test2@test.com",
+					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		if err := tc.query(tc.have, tc.options(tc.optionsArgs)); err != nil {
+			t.Error(err)
+		}
+
+		if !reflect.DeepEqual(tc.have, tc.want) {
+			t.Errorf("Error: expected %v, found %v", tc.want, tc.have)
+		}
+	}
+}
+
+func TestInsertQuery(t *testing.T) {
+	type queryFunc func(out interface{}, opts ...fury.QueryOption) error
+	type optionFunc func(conditions interface{}) fury.QueryOption
+
+	cases := []struct {
+		have interface{}
+		want interface{}
+	}{
+		{
+			&Account{
+				UserID: 3,
+			},
+			&Account{
+				UserID:    3,
+				Username:  "test3",
+				Password:  "test3",
+				Email:     "test3@test.com",
+				CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+				LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
+			},
+		},
+		{
+			&[]*Account{
+				&Account{
+					UserID: 4,
+				},
+				&Account{
+					UserID: 5,
+				},
+			},
+			&[]*Account{
```

Comment:
ini akan failed kan jika insert dengan berbeda type struct contoh:

```
// in top of the file
type Account2 struct {
	UserID    int `fury:"primary_key,auto_increment"`
	Username  string
	Password  string
	Email     string
	CreatedOn time.Time
	LastLogin time.Time
}
```

```
// lalu pada testcase nya di tambahkan 
&[]interface{}{
				&Account{
					UserID:    4,
					Username:  "test4",
					Password:  "test4",
					Email:     "test4@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
				&Account{
					UserID:    5,
					Username:  "test5",
					Password:  "test5",
					Email:     "test5@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
				&Account2{
					UserID:    5,
					Username:  "test5",
					Password:  "test5",
					Email:     "test5@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
````

Reply:
Ini karena design model-nya itu hanya untuk satu table. Jadi kalau dia terima selain pointer to struct atau pointer to slice dari pointer to struct akan error. Error message-nya dan comment-nya belum diganti, tulisannya masih `Error: expected pointer to struct, slice of pointer to struct or pointer to slice of pointer to struct, found`, harusnya cuma expect pointer to struct sama pointer to slice of pointer to struct aja

Reply:
sebaiknya jangan pakai type data `Interface{}` kosong untuk seperti ini, karena interface ini tidak miliki method dan akan menerima semua type data.  sebaiknya pakai `struct` atau inject nya ke setiap `Model`

contoh:

```
type Modeler interface{
  Insert()
  Update()
  Delete()
}
```

lalu implementasi nya
```
type Account struct {
	UserID    int `fury:"primary_key,auto_increment"`
         ................
}

func (a Account) Insert()
func (a Account) Delete()
func (a Account) Update()
```

Reply:
Hmm.. kalau seperti ini bukannya nanti waktu dipakai ngga bisa jadi semacam generic gitu? karena method-nya masih di-define di struct masing-masing?

Reply:
iyes, karena go type language jadi deklartif seperti ini dan  go memang belum support generic. dengan seperti ini kita dapat handle use case khusus seperti jika user ingin menambahkan pekerjaan kita perlu mencek apakah user tsb sudah membatasi limit penambahaan pekerjaan atau jika satu entity ingin melakukan transaction sebelum insert ke db.

Reply:
Bukannya kalau seperti ini kita violate SRP sama business logic-nya bisa kecampur sama repo ya?

Reply:
owh iya benar. sepertinya penjelasan saya diatas kurang benar. ini ada satu repo yang dapat kamu buat sebagai contoh dari maksud saya diatas. 

https://github.com/bxcodec/go-clean-arch/blob/master/author/repository.go

Reply:
Kalau begitu terpisah sih, untuk design-nya kemungkinan ada dua. Kalau di repo itu kan dia tidak menggunakan library untuk CRUD, melainkan langsung query. Cara lain kita bisa define method standar di `Modeler` yang akan diimplementasikan di struct repo-nya. Berarti yang jadi parameter-nya itu struct repo-nya? Atau bagaimana bagusnya ya?

Reply:
iya yang menjadi paramater-nya itu struct repo nya.

Reply:
Kalau misalkan seperti itu lebih baik untuk generate SQL bagaimana ya? Agak bingung untuk bagusnya di mana generator-nya ditaro dan apa saja kora-kira logic yang baik ditaro di setiap method `Insert`, dll.

Reply:
bisa sama saja kan seperti sekarang, tapi dari pada menggunakan empty interface pada  param insert model kamu dapat kamu dapat menggunakan interface dengan contract `insert`.

psudecode:
```
type ModelInserter interface{
func  Insert()
}

type ModelUpdater interface{
 func Update()
```

jika cara ini menurutmu kurang baik kamu dapat menggunakan cara lain yang lebih mudah untuk di implementasikan saat ini, karena approach lib kamu yang saat ini masih kurang baik, karena empty interface dapat menerima seluruh type data dan jika saya ingin menggunakan library ini, saya tidak bisa seacara intiutif bagaimana menggunakan library kamu tanpa membaca dokumentasi.